include/nbl/video/IGPUCommandPool.h

```diff
@@ -4,38 +4,835 @@
+                assert(false);
+                return nullptr;
+            }
+
+            CCommandSegment* nextSegment = new (nextSegmentMem) CCommandSegment;
+
+            cmdMem = nextSegment->allocate<Cmd>(args...);
+            if (!cmdMem)
+            {
+                assert(false);
+                return nullptr;
+            }
+
+            segmentListTail->setNext(nextSegment);
+            segmentListTail = segmentListTail->getNext();
+        }
+        Cmd* cmd = new (cmdMem) Cmd(std::forward<Args>(args)...);
+
+        if (segmentListHeadItr.m_cmd == nullptr)
+            segmentListHeadItr.m_cmd = cmd;
+
+        return cmd;
+    }
+
+    void deleteCommandSegmentList(CCommandSegment::Iterator& segmentListHeadItr, CCommandSegment*& segmentListTail)
+    {
+        auto& itr = segmentListHeadItr;
+
+        if (itr.m_segment && itr.m_cmd)
+        {
+            bool lastCmd = itr.m_cmd->getSize() == 0u;
+            while (!lastCmd)
+            {
+                ICommand* currCmd = itr.m_cmd;
+                CCommandSegment* currSegment = itr.m_segment;
+
+                itr.m_cmd = reinterpret_cast<ICommand*>(reinterpret_cast<uint8_t*>(itr.m_cmd) + currCmd->getSize());
+                currCmd->~ICommand();
+                // No need to deallocate currCmd because it has been allocated from the LinearAddressAllocator where deallocate is a No-OP and the memory will
+                // get reclaimed in ~LinearAddressAllocator
+
+                // We potentially continue to the next command segment under any one of the two conditions:
+                const bool potentiallyContinueToNextSegment =
+                    // 1. If the we run past the storage of the current segment.
+                    ((reinterpret_cast<uint8_t*>(itr.m_cmd) - itr.m_segment->getData()) >= IGPUCommandPool::CCommandSegment::STORAGE_SIZE)
+                    ||
+                    // 2. If we encounter a 0-sized command (terminating command) before running out of the current segment. This case will arise when the current
+                    // segment doesn't have enough storage to hold the next command.
+                    (itr.m_cmd->getSize() == 0);
+
+                if (potentiallyContinueToNextSegment)
+                {
+                    CCommandSegment* nextSegment = currSegment->getNext();
+
+                    currSegment->~CCommandSegment();
+                    m_commandSegmentPool.deallocate(currSegment, COMMAND_SEGMENT_SIZE);
+
+                    if (!nextSegment)
+                        break;
+
+                    itr.m_segment = nextSegment;
+                    itr.m_cmd = itr.m_segment->getFirstCommand();
+                }
+
+                lastCmd = itr.m_cmd->getSize() == 0u;
+                if (lastCmd)
+                {
+                    currSegment->~CCommandSegment();
+                    m_commandSegmentPool.deallocate(currSegment, COMMAND_SEGMENT_SIZE);
+                }
+            }
```

Comment:
this whole pile of code should just be
```cpp
if (segmentListTail)
{
   CCommandSegment* prevSegment = segmentListHeadItr.m_segment;
   assert(segmentListHeadItr.m_cmd && prevSegment);
   for (auto& itr=segmentListHeadItr; itr!=segmentListTail->end(); )
   {
      ICommand* cmd = *(itr++);
      cmd->~ICommand();
      // No need to deallocate currCmd because:
      // - it has been allocated from the LinearAddressAllocator where deallocate is a No-OP
      // - the memory will get reclaimed in `~LinearAddressAllocator` when whole segment is destroyed
      if (itr.m_segment==lastSegment)
         continue;

      // segment has changed, we need to free the last one
      prevSegment->~CCommandSegment();
      m_commandSegmentPool.deallocate(lastSegment, COMMAND_SEGMENT_SIZE);
   }
   // the last segment is always freed outside the loop
   assert(segmentListTail==prevSegment);
   segmentListTail->~CCommandSegment();
   m_commandSegmentPool.deallocate(segmentListTail, COMMAND_SEGMENT_SIZE);
}
else
{
   assert(!segmentListHeadItr.m_cmd && !segmentListHeadItr.m_segment);
}
```

Reply:
Resolved.